components/brave_shields/browser/site_specific_script_service.cc

```diff
@@ -0,0 +1,142 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_shields/browser/site_specific_script_service.h"
+
+#include <algorithm>
+#include <utility>
+
+#include "base/base_paths.h"
+#include "base/bind.h"
+#include "base/files/file_util.h"
+#include "base/json/json_reader.h"
+#include "base/logging.h"
+#include "base/macros.h"
+#include "base/memory/ptr_util.h"
+#include "base/strings/utf_string_conversions.h"
+#include "base/threading/thread_restrictions.h"
+#include "brave/browser/brave_browser_process_impl.h"
+#include "brave/components/brave_shields/browser/ad_block_service.h"
+#include "brave/components/brave_shields/browser/local_data_files_service.h"
+#include "brave/components/brave_shields/browser/dat_file_util.h"
+
+namespace brave_shields {
+
+SiteSpecificScriptService::SiteSpecificScriptService()
+    : weak_factory_(this) {
+  DETACH_FROM_SEQUENCE(sequence_checker_);
+}
+
+SiteSpecificScriptService::~SiteSpecificScriptService() {
+}
+
+SiteSpecificScriptService::SiteSpecificScriptRule::SiteSpecificScriptRule() =
+  default;
+SiteSpecificScriptService::SiteSpecificScriptRule::SiteSpecificScriptRule(
+  const SiteSpecificScriptRule& other) = default;
+SiteSpecificScriptService::SiteSpecificScriptRule::~SiteSpecificScriptRule() =
+  default;
+
+bool SiteSpecificScriptService::ScriptsFor(const GURL& primary_url,
+                                           std::vector<std::string>* scripts) {
+  bool any = false;
+  scripts->clear();
+  for (const auto& rule : rules_) {
+    for (const auto& pattern : rule.urls) {
+      if (pattern.MatchesURL(primary_url)) {
+        for (const auto& path : rule.scripts) {
+          std::string contents;
+          if (base::ReadFileToString(path, &contents)) {
```

Comment:
I think this is causing file IO on the main thread when pages load so I think this needs to happen somewhere else. I think we have so few scripts that we can just read them once into memory after startup.

Reply:
this would be a perf bottleneck anyway. I would suggest looking at how content scripts are loaded